### ISSUE TITLE
Merge tagger binary formats into a single format

### ISSUE THREAD

**Issue by ftyers** [apertium/apertium#89]
We have 5 (or more?) different binary formats for the tagger that are supported. At least:

* Unigram model 1
* Unigram model 2
* Unigram model 3
* 2-gram HMM 
* Perceptron
And additionally two more that are not supported:
* 3-gram HMM
* LSWpost

Having different binary formats isn't an enormous problem (although for the unigram tagger it seems excessive) but it seems that having the `apertium-tagger` program decide which one it is reading is better than having the user decide (see e.g. https://github.com/apertium/apertium-oci/issues/6).

**Comment by TinoDidriksen** [comment]
Linked issue: https://github.com/apertium/apertium-init/issues/31

**Comment by flammie** [comment]
+1 on creating single binary format with enough metadata to decide what to do with the file, it's quite simple to do imo and improves usability a lot.

**Comment by mr-martian** [comment]
Is there documentation somewhere of what the different formats are currently?

**Comment by mr-martian** [comment]
In answer to my own question, here's the current binary formats:

## Unigram Model 1

Uses `serialiser.h`

Data structure:
```
map<Analysis, size_t> Model1
  analyzed form => number of occurences in training data
```
Data types:
```
Analysis:
  vector<Morpheme>  # compound parts

Morpheme:
  wstring           # lemma
  vector<Tag>       # tags

Tag:
  wstring           # tag name
```

## Unigram Model 2

Uses `serialiser.h`

Data structure:
```
map<a, map<Lemma, size_t>> Model2
  analyzed form without leading lemma => leading lemma => number of occurrences
```
Data types:
```
a:
  vector<Tag>       # tags of first compound part
  vector<Morpheme>  # subsequent compound parts

Lemma:
  wstring           # lemma

Morpheme:
  wstring           # lemma
  vector<Tag>       # tags

Tag:
  wstring           # tag name
```

## Unigram Model 3

Uses `serialiser.h`

Data structure:
```
map<i, map<Lemma, size_t>> Model3_l_t
  tags of first compound part => leading lemma => count
map<i, map<Lemma, size_t>> Model3_cl_ct
  tags of nth compount part => lemma of n+1th compound part => count
map<Lemma, map<i, size_t>> Model3_ct_cl
  lemma of nth compound part => tags of nth compound part => count
```

Data types:
```
i:
  vector<Tag>

Lemma:
  wstring
```

## HMM

Uses `compression.h`

Data structure:
```
set<TTag> open_class
vector<TForbidRule> forbid_rules
vector<wstring> array_tags
map<wstring, TTag> tag_index
vector<TEnforceAfterRule> enforce_rules
vector<wstring> prefer_rules
ConstantManager constants
Collection output
double[N][N] a
double[N][M] b
PatternList plist
vector<wstring> discard
```

Data types:
```
TTag:
  int

TForbidRule:
  TTag
  TTag

TEnforceAfterRule:
  TTag
  vector<TTag>

ConstantManager:
  map<wstring, int>

Collection:
  an indexed set of set<int>

PatternList:
  (lttoolbox object)
```

Format notes:

`open_class` writes differences between tags rather than the tags themselves

`a` and `b` are written as
```
N
M
all entries of a using endian_double_util.h
entries of b which appear in output in the form [ index1 index2 value ]
  also written with endian_double_util.h
```

`discard` is only written if it is non-empty

## Perceptron

Uses `serialiser.h`

Data structure:
```
PerceptronSpec spec
FeatureVec weights
```

Data types:
```
PerceptronSpec:
  int beam_width
  vector<string> str_consts
  vector<set<string>> set_consts
  vector<FeatureDefn> features
  vector<FeatureDefn> global_defns
  FeatureDefn global_pred
  Optional<TaggerDataPercepCoarseTags> coarse_tags

FeatureVec:
  map<vector<string>, double>

FeatureDefn:
  vector<unsigned char>

TaggerDataPercepCoarseTags:
  set<TTag> open_class
  vector<wstring> array_tags
  map<wstring, TTag> tag_index
  ConstantManager constants
  Collection output
  PatternList plist
```

Format notes:

`coarse_tags` is written as `1` and then the value if present, or just `0` if not.

## LSW

Uses `compression.h`

LSW has essentially the same data structure as HMM except with `double[N][N][N] d` in place of the `a` and `b` matrices, which is written as non-empty entries as `b` is.

**Comment by mr-martian** [comment]
As far as I can tell, `serialiser.h` was added in the process of implementing the unigram taggers and is only used for them and the perceptron tagger.

**Comment by mr-martian** [comment]
Will it be necessary to provide a way of converting from the old binary format to the new one or can we assume that retraining on the same corpora will be sufficient?

**Comment by TinoDidriksen** [comment]
I would prefer if current files keep working, but new ones are given a header that identifies them and future-proofs them, a'la https://github.com/apertium/lttoolbox/blob/master/lttoolbox/compression.h#L28

**Comment by ftyers** [comment]
For the unigram tagger, I think that retraining is fine. For the perceptron and bigram HMM taggers, retraining those `.prob` files might be a challenge. Although honestly, we should get better and making this stuff less brittle.

**Comment by mr-martian** [comment]
https://gist.github.com/mr-martian/a8d1562ee95a0f636effd37b35e33171

The above is a python implementation of our 2.5 methods of reading and writing binary data (3 for floats and 2 for everything else) and I could probably turn this into a general conversion script without too much trouble.

**Comment by TinoDidriksen** [comment]
Any new format should be mmap-compatible so that we can eventually avoid loading it into memory at all - so no packing of values, and keep it little-endian on disk.

The way `compression.h` packs integers made sense 20 years ago, but stopped making sense 10 years ago.

**Comment by mr-martian** [comment]
If `modes.xml` is to be trusted, Perceptron is only used by English, and Unigram 2 is used by the following places:

```
./apertium-nhi-nhn
./apertium-oci
./apertium-tur-tat
./apertium-tur-uzb
./apertium-nci-nhi
./apertium-fao-nor
./apertium-hin-pan
./apertium-kan-mar
./apertium-uig-tur
./apertium-oci-fra
./apertium-grn-spa
./apertium-crh-tur
./apertium-tur-tki
./apertium-spa-ote
./apertium-scn-spa
./apertium-scn
./apertium-kaz-sah
./apertium-quc-spa
./apertium-oci-cat
./apertium-hin
./apertium-tat-eng
```

Everything else has no option, and so is presumably HMM bigram.

And I guessed that there were a handful of `.prob` files that got copied a lot, so I spent rather longer than I meant to determining that there are 103 distinct `.prob` files across all the repos in the org (398 total).

**Comment by xavivars** [comment]
Regarding perceptron being only in English, I remember when that was trained (during the en-cat GSoC) it was easier to train that the HMM (due to lack of documentation).

I completely agree that having all taggers in the same binary format would be better, I would actually like to know if there's a real performance difference between the multiple taggers.

**Comment by mr-martian** [comment]
I'm currently working on this in #130 

My current design plan is as follows (feedback welcome):

The new binary format will start with a header like the transducer one, probably `APTG` and then feature flags, for the moment just specifying which of the 6 taggers it is (the code for LSW exists and is close enough to HMM that it seems like I might as well include it). If `apertium-tagger` doesn't find that header, it will successively try to load the file as each type, probably in the order HMM, Unigram 2, Perceptron, the other 3.

Having anything as a purely binary format feels very ugly to me, so I also intend to create a simple text format containing the same data, which I think will solve several issues, such as the lack of documentation of training settings and apertium-init's tendency to generate HMM taggers and then overwrite them with empty Unigram 2 taggers the first time the user runs `make`.

Since the taggers aren't retrained very often, it seems to me that complete backwards compatibility of that interface is less important and so I'm thinking of having the tagger trainer operate entirely on text files, and maybe be a separate command-line program (so if you run `apertium-tagger -s` it will tell you to convert your `.prob` file with `apertium-convert-tagger` and then run `apertium-train-tagger`). This also allows me to put a lot of the deprecated functions for reading binary data in a single place and very clearly label them as deprecated.

**Comment by mr-martian** [comment]
New proposal:

All of the tagger models are equivalent to single-layer perceptrons with various restrictions on what the features can be.

Thus I would like to amend my previous plan and in fact have the new binary form not specify what model it is but rather have there be only 1 model with compatibility code which reinterprets the existing files as perceptrons.

**Comment by mr-martian** [comment]
Yes, old `.prob` files would continue working, but instead of their being an HMM tagger it would be because file reading code interprets HMM files as Perceptrons where the features are mutually exclusive, the context size is 1, and the beam search has size 1.

**Comment by mr-martian** [comment]
```
Parameters:
window width W
depth D
beam search size B

Algorithm:
for word in input:
    for reading in word:
        features = [extract from an FST] + [last W selected readings] + [next W surface forms]
        weight = sum of lookup(group) for group in [all size D subsets of features]
    select max(weight), using beam search of size B
```

| Model | `B` | `D` | `W` |
|----------|-----|------|-----|
| Unigram | 1 | 1 | 0 |
| HMM | 1 | 2 | 1 |
| LSW | 1 | 3 | 1 |
| English Perceptron | 5? | 2 | 2 |

In all of these except the English Perceptron, the features have been constructed such that a given reading will match at most one of them, but the algorithm above does not care.

LSW is the only one that looks at tags on a following word, which makes things a little more complicated, but not too bad. (Alternatively, I see no evidence that anyone has ever actually used LSW, so if dropping it simplifies things it might not be a bad idea.)

In theory an HMM tagger would be a bit more complicated, but the one in `apertium-tagger` doesn't do any backtracking and just selects the current highest-probability option.

So really this would just be simplifying what the developer has to consider at training time (or not, the training code could probably be made to mimic the restrictions decently well).

**Comment by xavivars** [comment]
LSW [was implemented as a GSoC project](https://wiki.apertium.org/wiki/User:Gang_Chen), ([paper here](https://arxiv.org/abs/1509.05517), [based on this other paper from @mlforcada](https://www.dlsi.ua.es/~mlf/docum/sanchezvillamil04p.pdf)), so I'm not sure if I'd remove it... 

But it's also true that if no-one has use it since 2014, and it performs similarly (or even worse) than the HMM (as mentioned in those papers), the loss won't be that bad, and the code will always be in GIT in case anyone wants to recover it...
